### ISSUE TITLE
feat: Export Chart Relationships to Graphviz DOT Format

### ISSUE THREAD

**Issue by mits87** [kreuzberg-dev/kreuzberg#579]
### Summary

Add support for extracting chart/diagram structures (nodes, edges, labels, and styling) into the [Graphviz DOT format](https://graphviz.org/doc/info/lang.html)￼ - [graphiz_rust](https://docs.rs/graphviz-rust/latest/graphviz_rust/). 

This would allow users to programmatically reconstruct diagrams, visualize relationships, and integrate with graph-based tooling.

### Problem

Currently, when processing documents with diagrams (e.g. org charts, flowcharts, architecture diagrams), the extracted output lacks a structured representation of relationships between elements.

This makes it difficult to:
* Reconstruct diagrams programmatically
* Analyze relationships between entities
* Visualize or export diagrams into other systems
* Perform graph-based reasoning or transformations

### Proposed Solution

Extend the extraction pipeline to detect and output diagram structures in DOT format.

The system should:

1. Detect Nodes
* Identify all diagram elements (boxes, shapes, entities) and node colors (if applicable).
* Extract texts (eg. using OCR)

2. Detect Edges (Connections)
* Identify relationships between nodes (arrows, lines)
* Extract:
  * Source node
  * Target node
  * Direction (if applicable)
  * Edge labels (if present)

3. Capture Styling (Optional but valuable)
* Node color / fill
* Border styles
* Edge styles (dashed, solid, etc.)

4. Preserve Text Context
* Text inside nodes
* Text between nodes (e.g. relationship labels)

### Expected Behavior

When a document contains a diagram:
* The system outputs a DOT representation alongside existing extraction results

The output should be:
* Deterministic
* Clean (no duplicates)
* Referencing consistent node IDs

### Use Cases
* Org chart extraction
* System architecture diagrams
* Process flows
* Dependency graphs
* Legal/entity relationship mapping

### Why This Matters

This unlocks:
* Graph-based processing (Neo4j, NetworkX, etc.)
* Visualization pipelines (Graphviz, D3, Mermaid conversion)
* Better explainability (traceable relationships vs flat text)
* Stronger downstream automation (agents operating on graphs)

### Example diagram output

```
digraph G {
  rankdir=TB;

  node1 [label="Node A" shape=box fillcolor="#a6cee3"];
  node2 [label="Node B" shape=diamond fillcolor="#b2df8a"];
  node3 [label="Node C" shape=box fillcolor="#fb9a99"];
  node4 [label="Node D" shape=ellipse fillcolor="#fdbf6f"];

  note1 [label="Some annotation" shape=note];

  node1 -> node2 [label="50%"];
  node2 -> node3 [label="relates"];
  node3 -> node4;
  node2 -> note1 [style=dashed arrowhead=none];
}
```

### Scope / Notes
* Should support extraction from both PDFs and images
* The diagram output should be an additional structured extraction, included under content.diagram (without affecting existing outputs)
* All detected nodes and edges must be traceable to source, including at minimum page reference and bounding box (for validation and debugging)

### Examples

<img width="2000" height="1500" alt="Image" src="https://github.com/user-attachments/assets/f1a92237-cfb9-481f-ac0b-311df5fb8e1f" />

<img width="1337" height="708" alt="Image" src="https://github.com/user-attachments/assets/ec97a678-ac75-4d78-98ef-cdeb7c05fa34" />

<img width="1536" height="1024" alt="Image" src="https://github.com/user-attachments/assets/30d03f4e-988b-4a13-b7f8-e50f31fd4938" />

<img width="1252" height="929" alt="Image" src="https://github.com/user-attachments/assets/27a34e0f-8975-40d8-b49c-112d8e039146" />

### Observation

Most of the Kreuzberg pipeline is there (rasterize pdf pages, OCR and text recognition, text coordinates etc.), the missing part is diagram / charts recognition (nodes and edges + text connection).

**Comment by Goldziher** [comment]
Thanks for the detailed feature request — this is a compelling use case and we'd love to support it.

We've done extensive research into what's available today for diagram element detection (nodes, edges/arrows, labels) and want to share our findings transparently.

## Current State of Available Models

**The core challenge is Stage 2 — parsing the internals of a diagram.** Kreuzberg already handles Stage 1 (detecting diagram/figure regions in documents via our layout detection models) and Stage 3 (OCR for text within detected regions). What's missing is a model that can detect individual shapes (boxes, diamonds, ovals), arrows/connections, and their relationships.

Here's what we found:

- **No permissively-licensed ONNX model exists** that detects diagram elements (nodes, arrows, shapes) at production quality. The closest candidate (ELCA-SA/BPMN_Detection, F1=0.96 objects / 0.92 arrows) is MPL-2.0 licensed, which is incompatible with our codebase.
- **Grounding DINO** (Apache 2.0, ONNX available) can do zero-shot text-prompted detection ("arrow", "rectangle", "diamond"), but it was trained entirely on photographs — performance on technical diagrams is untested and likely insufficient.
- **google/deplot** (Apache 2.0) only handles data charts (bar, line, pie) → table conversion. It cannot process flowcharts or relational diagrams.
- **diagram2graph** (Apache 2.0, Qwen2.5-VL-3B) does end-to-end diagram → JSON extraction but is a 3B parameter VLM — far too large for ONNX deployment in our pipeline.

## Permissive Datasets That Exist

The training data situation is better:
- **hdBPMN** (CC-BY 4.0) — 502 hand-drawn BPMN diagrams, 25 element types, COCO format
- **FloCo** (MIT) — 11,884 flowchart images
- **Roboflow flowchart datasets** (CC-BY 4.0) — ~774 annotated images
- **DIDI** (Google Research) — 58K+ digital ink diagrams

## Our Plan

We're adding this to our roadmap as a later-phase feature. The path forward is:

1. **Fine-tune our existing RT-DETRv2 or YOLO model** on the permissive datasets above, with an added keypoint head for arrow start/end detection
2. **Export to ONNX** and integrate into the existing kreuzberg layout detection pipeline as a second-stage model
3. **Build the graph construction** post-processing: match arrow endpoints to node bounding boxes, OCR text within nodes, emit DOT/JSON output

This keeps everything MIT-licensed and within our existing ONNX inference pipeline. We'll update this issue as we make progress.

The observation in the issue is exactly right — most of the pipeline is already there. The missing piece is the diagram element detection model, and we'll need to train one ourselves on permissive data.

**Comment by mits87** [comment]
@Goldziher Hi! Just following up on this feature request — do you have any rough timeline for when this could be implemented?

This functionality is quite important on our side, so it would really help us plan accordingly if there’s any estimate (even a rough one).

If this isn’t something that can be delivered in the near term, would it be possible to expose at least a lightweight signal indicating whether a document likely contains a diagram / org chart / ownership structure? For example:

```
{
  "diagram": {
    "confidence": 0.8
  }
}
```

Even a simple confidence score (or boolean flag) per chunk / page would already allow us to improve routing and handling on our side.

Thanks a lot!

**Comment by Goldziher** [comment]
@mits87 -- so realilstically this will probably have to be done via hybrid extraction. Namely, we will need to extract diagrams and send them into VLM extraction. It will be included in our commercial offering for certain (our SAAS / self-hosted solution). If this is interesting for you guys, lets have a chat (via either discord or email -- we can set a meeting as wel). 

For extracting confidence info -- this is doable now. I'll try to make it in the next minor version. Which will come next week or even earlier.

**Comment by Goldziher** [comment]
@mits87 -- api is extended, and in the next verison you will be able to get the data. BUT the layout model does not distinguish between "picture" and "diagram" - its all "picture".

**Comment by mits87** [comment]
@Goldziher this is already very helpful, thanks!

One follow-up: would it be possible to add an option to plug in a custom detection mechanism (e.g. a vision LLM classifier), so I can determine what’s on the page on my side?

**Comment by Goldziher** [comment]
> @Goldziher this is already very helpful, thanks!
> 
> One follow-up: would it be possible to add an option to plug in a custom detection mechanism (e.g. a vision LLM classifier), so I can determine what’s on the page on my side?

Open a separate issue please

**Comment by Goldziher** [comment]
@MannXo wanna take a crack at this issue? or more precisely, see if you can find some solution for this?

**Comment by MannXo** [comment]
@Goldziher sure thing

**Comment by MannXo** [comment]
Had a look. One angle that may be missing above: the analysis assumes diagrams arrive as pixels, but many org charts and architecture diagrams are vector, and there the graph is already in the file.

Quick proof against `test_documents/xml/org_chart.svg`, which we already ship:

```
9 nodes, 3 edges
n0 [label="Jane Smith / Chief Executive Officer" fillcolor="#2c3e50"];
n1 [label="Bob Chen / Chief Technology Officer" fillcolor="#2980b9"];
...
n0 -> n1;  n0 -> n2;  n0 -> n3;
```

Every node, label, fill colour and edge direction correct. No model, no new dependency, about 100 lines. `resvg`/`usvg` are already dependencies and `core/image_encode.rs` already parses SVG into a full `usvg::Tree` before discarding the structure to rasterize.

Same applies to vector PDFs. `pdf_oxide` exposes `extract_paths_in_rect(page, region)`, which we already call in `pdf/layout_gate.rs`, and `PathContent` carries `fill_color`, `stroke_color`, `stroke_width` and `dash_pattern`. That is the entire "Capture Styling" section, losslessly, which detection cannot recover from pixels. It also composes with the layout detection we have: stage 1 finds the figure region, this reads inside it. PPTX would be more exact still and we do not parse it at all today, since `<p:cxnSp>` connectors carry `stCxnId`/`endCxnId`.

This does nothing for raster diagrams, so the model plan stands there. But it would cover a fair share of real documents deterministically, which is what the issue asks for under Expected Behavior.

@Goldziher  Three questions before I build anything:

1. Do you want this in OSS, given the VLM hybrid is going into the commercial offering?
2. Where should DOT output live, a renderer like `doctags`, a field on the result, or something else?
3. How far should a first PR go, SVG only or vector PDF as well?

One implementation detail either way: usvg is compiled without its `text` feature here, so node labels need either that feature enabled or our own text layer.

**Comment by Goldziher** [comment]
> Had a look. One angle that may be missing above: the analysis assumes diagrams arrive as pixels, but many org charts and architecture diagrams are vector, and there the graph is already in the file.
> 
> Quick proof against `test_documents/xml/org_chart.svg`, which we already ship:
> 
> ```
> 9 nodes, 3 edges
> n0 [label="Jane Smith / Chief Executive Officer" fillcolor="#2c3e50"];
> n1 [label="Bob Chen / Chief Technology Officer" fillcolor="#2980b9"];
> ...
> n0 -> n1;  n0 -> n2;  n0 -> n3;
> ```
> 
> Every node, label, fill colour and edge direction correct. No model, no new dependency, about 100 lines. `resvg`/`usvg` are already dependencies and `core/image_encode.rs` already parses SVG into a full `usvg::Tree` before discarding the structure to rasterize.
> 
> Same applies to vector PDFs. `pdf_oxide` exposes `extract_paths_in_rect(page, region)`, which we already call in `pdf/layout_gate.rs`, and `PathContent` carries `fill_color`, `stroke_color`, `stroke_width` and `dash_pattern`. That is the entire "Capture Styling" section, losslessly, which detection cannot recover from pixels. It also composes with the layout detection we have: stage 1 finds the figure region, this reads inside it. PPTX would be more exact still and we do not parse it at all today, since `<p:cxnSp>` connectors carry `stCxnId`/`endCxnId`.
> 
> This does nothing for raster diagrams, so the model plan stands there. But it would cover a fair share of real documents deterministically, which is what the issue asks for under Expected Behavior.
> 
> @Goldziher  Three questions before I build anything:
> 
> 1. Do you want this in OSS, given the VLM hybrid is going into the commercial offering?
> 2. Where should DOT output live, a renderer like `doctags`, a field on the result, or something else?
> 3. How far should a first PR go, SVG only or vector PDF as well?
> 
> One implementation detail either way: usvg is compiled without its `text` feature here, so node labels need either that feature enabled or our own text layer.
> 

The OSS is the primitive layer. Hybrid extraction is indeed out of scope here. Everything else is in scope - so go ahead

**Comment by MannXo** [comment]
Tracking comment for the vector-diagram work in #1410.

**Done**

- `dot` output format, reachable via `OutputFormat::Custom("dot")` through both `FromStr` and serde, so no binding regeneration.
- Format-independent recovery: `DiagramGraph` types, and an `assemble` step that turns outlines, connectors and labels into nodes and edges. Node and edge matching is written once and shared.
- SVG front end. Geometry from `usvg`, labels from a second pass over the source XML, since `usvg` is built here without its `text` feature and drops text during conversion.
- Vector PDF front end via `pdf_oxide`, producing the same intermediates.
- Scoring against the diagram corpus, driven by `diagrams/manifest.json` so fixtures added upstream are measured without a change here.

**Where recovery stands**, on the 26-fixture corpus: 24 at 1.0 for both node and edge recall, including Graphviz (clusters, records, self-loops, orthogonal routing, CJK, a 128-node graph), Mermaid, PlantUML and LibreOffice, with all six negative controls correctly recovering no graph at all.

Every `_geometry` variant scores exactly what its original scores, which is what says recovery reads coordinates rather than the graph each producer writes back into its own output.

**Not there yet**, pinned at their current scores so the gap stays visible and a regression still fails:

- `graphviz_large`, 137 of 141 edges. The four are long-range shortcuts that pass close to intervening nodes.
- `icon_nodes`, 0 of 4. The AWS/Azure architecture style draws a node as an icon glyph with no enclosing outline and a caption underneath, which needs both merging a glyph's several subpaths into one node and reading a caption that sits below a shape. The second is delicate: a caption below a shape and an edge label above the next one are the same geometry.

**Blocked upstream:** PDF recovery currently loses one arrowhead per drawing and misattributes fills on filled-and-stroked shapes, because `extract_paths` in pdf_oxide never finalizes a path painted with `B`, `B*` or `b*`. Filed as yfedoseev/pdf_oxide#1028 with a repro and the fix. Node recovery from PDF is unaffected and exact.

**Out of scope here**, per the discussion above: hybrid extraction. Still open from the review: an OSS-tool comparison, benchmark-harness integration and profiling.

**Comment by Goldziher** [comment]
First part is done. The remainder requires dedicated models etc. So keeping it open for now. 

@MannXo and @tobocop2 - please check what can be done regarding models.

**Comment by MannXo** [comment]
One result from #1410 shapes the options. `assemble` is format independent. It takes outlines,
connectors and labels and produces the graph, so SVG and PDF are only front ends. A model does
not need to output a graph. It needs to output shapes, lines and text, and it becomes a third
front end. Arrowhead direction, container and table rejection, node ordering, dedup and the DOT
renderer are already there and tested.

That split also protects determinism. A VLM emitting DOT directly cannot give it and will invent
edges that are not drawn, which is the worst failure mode here because a plausible wrong edge
looks exactly like a right one downstream. I would keep a VLM as a last tier behind a flag, not
as the default path.

Evaluation before architecture. Rasterising the corpus fixtures gives a raster benchmark whose
ground truth is already written and already scored by the harness in #1410. Cheap, useful
whichever approach wins, and without it there is no honest way to compare two.

Edges are the open question. Nodes are compact rectangles, which detectors handle well. Edges are
thin, long, often curved and overlapping, and a box around a diagonal arrow says almost nothing.
Line segment detection, a relation head over node pairs, or something else is what I want to
settle on the eval set rather than argue in advance.

@tobocop2 how do you want to split this? I can take the raster eval set and the survey of what is
available, happy to swap if you are already on one.

Two things worth deciding early:

1. New weights in `xberg-io/layout-models` or a separate repo?
2. Is a remote VLM acceptable as a fallback tier at all, or should this stay fully local?

**Comment by MannXo** [comment]
Raster eval set is up as xberg-io/test_documents#3, 12 PNG fixtures against the ground truth that already exists, so SVG, vector PDF and raster score apart. Needs publishing before it can be used.
FYI @Goldziher

**Comment by tobocop2** [comment]
@MannXo feel free to take both the eval set and the survey. I will review.

**Comment by Goldziher** [comment]
Shipped. Implemented in #1410 (*feat(diagram): recover Graphviz DOT from vector SVG diagrams*, merged 2026-08-10) and documented in `05d3c196a7`. Ships in the upcoming **v1.1.0**.

Diagram structure is recovered from vector SVG — shapes become nodes, connectors become edges, labels are attached to the shape that encloses them, and the result is emitted as Graphviz DOT (`crates/xberg/src/rendering/dot.rs`), reachable via a custom output format.

@mits87 — thank you for the original request and for the DOT-language and `graphviz-rust` references; the emitted syntax follows them.

Two follow-ups from the same area are tracked separately rather than left implicit here:

- **#1420** — chart gridlines that cross the bars were being read as edges, so a bar chart recovered as a graph. Fix in flight: gridline families are now identified structurally (≥3 same-style axis-aligned strokes sharing a run-span at regular offsets) and dropped before edge building, rather than by an endpoint-distance threshold, which the negative controls showed does not generalise.
- Clusters posing as nodes, and icon captions, were fixed in `5af8a0e7e5`.
